File: packages/b2c-tooling-sdk/src/operations/sites/sites-backend.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
import type {SitesBackend} from './sites-types.js';
import {OcapiSitesBackend} from './ocapi-sites-backend.js';
import {ScapiSitesBackend} from './scapi-sites-backend.js';
import {createDualBackend, type DualBackendConfig} from '../../clients/dual-backend-factory.js';

export type SitesBackendConfig = DualBackendConfig;

/**
 * Builds a Sites backend for read operations (list/get). In `auto` mode
 * (the default) it prefers SCAPI (`site/sites/v1`) and falls back to the
 * deprecated OCAPI Data API on `invalid_scope`.
 */
export function createSitesBackend(config: SitesBackendConfig): SitesBackend {
  return createDualBackend<SitesBackend>(config, {
    domainName: 'Sites',
    Scapi: ScapiSitesBackend,
    Ocapi: OcapiSitesBackend,
  });
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/operations/sites/sites_backend.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Factory selecting between the SCAPI and OCAPI sites backends.

Mirrors ``src/operations/sites/sites-backend.ts``.
"""

from __future__ import annotations

from b2c_tooling_sdk.clients.dual_backend_factory import DualBackendConfig, DualBackendCtors, create_dual_backend
from b2c_tooling_sdk.operations.sites.ocapi_sites_backend import OcapiSitesBackend
from b2c_tooling_sdk.operations.sites.scapi_sites_backend import ScapiSitesBackend
from b2c_tooling_sdk.operations.sites.sites_types import SitesBackend

#: Configuration for :func:`create_sites_backend`. Alias of the shared
#: :class:`~b2c_tooling_sdk.clients.dual_backend_factory.DualBackendConfig`.
SitesBackendConfig = DualBackendConfig


def create_sites_backend(config: SitesBackendConfig) -> SitesBackend:
    """Builds a Sites backend for site and cartridge-path operations.

    In ``auto`` mode (the default) it prefers SCAPI (``site/sites/v1``) and
    falls back to the deprecated OCAPI Data API on a safe capability/auth/request
    rejection.
    """
    return create_dual_backend(
        config,
        DualBackendCtors[SitesBackend](
            domain_name="Sites",
            scapi=lambda cfg: ScapiSitesBackend(cfg),
            ocapi=lambda inst: OcapiSitesBackend(inst),
        ),
    )


__all__ = ["SitesBackendConfig", "create_sites_backend"]
